File: src/ros2_bridge/mappers/geometry_msgs/inertia.rs
//! Typed mapper for `geometry_msgs/msg/Inertia`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn inertia_to_bus(
    msg: ros_env::geometry_msgs::msg::Inertia,
) -> crate::geometry_msgs::msg::v1::Inertia {
    crate::geometry_msgs::msg::v1::Inertia {
        m: msg.m,
        com: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.com)),
        ixx: msg.ixx,
        ixy: msg.ixy,
        ixz: msg.ixz,
        iyy: msg.iyy,
        iyz: msg.iyz,
        izz: msg.izz,
    }
}

pub(crate) fn inertia_to_ros(
    bus: crate::geometry_msgs::msg::v1::Inertia,
) -> ros_env::geometry_msgs::msg::Inertia {
    ros_env::geometry_msgs::msg::Inertia {
        m: bus.m,
        com: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(
            bus.com.unwrap_or_default(),
        ),
        ixx: bus.ixx,
        ixy: bus.ixy,
        ixz: bus.ixz,
        iyy: bus.iyy,
        iyz: bus.iyz,
        izz: bus.izz,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsInertiaMapper;

impl TypedTopicMapper for GeometryMsgsInertiaMapper {
    type Ros = ros_env::geometry_msgs::msg::Inertia;
    type Bus = crate::geometry_msgs::msg::v1::Inertia;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(inertia_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(inertia_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/inertia.py
"""Generated mapper for `geometry_msgs/msg/Inertia`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.vector3 import vector3_to_bus, vector3_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import Inertia as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def inertia_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.m = msg.m
    bus.com.CopyFrom(vector3_to_bus(msg.com))
    bus.ixx = msg.ixx
    bus.ixy = msg.ixy
    bus.ixz = msg.ixz
    bus.iyy = msg.iyy
    bus.iyz = msg.iyz
    bus.izz = msg.izz
    return bus


def inertia_to_ros(bus):
    from geometry_msgs.msg import Inertia as RosMsg

    out = RosMsg()
    out.m = bus.m
    out.com = vector3_to_ros(bus.com)
    out.ixx = bus.ixx
    out.ixy = bus.ixy
    out.ixz = bus.ixz
    out.iyy = bus.iyy
    out.iyz = bus.iyz
    out.izz = bus.izz
    return out


class GeometryMsgsInertiaMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import Inertia as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return inertia_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return inertia_to_ros(bus)
